### PR TITLE
feat(scripts): extract session-state.json surgical updater to session-state.sh (#284 C-11)

### LINKED ISSUE #284: Extract session-state.json surgical updater to .claude/scripts/session-state.sh (P1, #271 C-11)
**Parent audit:** #271 (audit landed in PR #272)
**Audit doc:** `.claude/reference/script-extraction-audit.md` row **C-11**
**Priority:** P1## Summary

Read-modify-write `~/.claude/session-state.json` with sibling-field preservation and atomic replace. Eliminates the verbose inline jq blocks that risk wiping other top-level fields if written naively.

## Where it lives today

`phase-b-reviewer` Greptile budget check (×2 inline jq expressions), `continue` Step 6 rate-limit branch, `handoff-files.md` reference, `pm` resume mode.

## Why extract

- **Call sites:** 2 agents + 1 skill + 1 reference
- **Tokens saved per session (est):** ~450
- **Determinism:** Fully deterministic
- **Risk:** Medium — atomic-write must match the pattern in `repair-trust-all.sh`

## Proposed CLI signature

```
session-state.sh --get <jq-path>
session-state.sh --set <jq-path>=<value> [--set ...]
```

Multiple `--set` flags merge into one atomic write.

## Exit-code contract

| Code | Meaning |
|------|---------|
| 0 | OK |
| 2 | Usage error |
| 3 | State file missing (creates default on `--set`, errors on `--get`) |
| 4 | jq parse error |
| 5 | Write failed |

## Suggested approach

Use the tempfile + `mv` atomic-write pattern from `.claude/scripts/repair-trust-all.sh`. Use `jq --arg` for typed values; never string-concat into `jq` expressions.

## Related

- Parent audit: #271
- Audit PR: #272
- C-12 (Greptile budget) is the primary consumer

## Acceptance criteria

- [ ] Script created at `.claude/scripts/session-state.sh` with the CLI signature above
- [ ] Exit codes match the contract above
- [ ] `session-state.sh --help` prints usage
- [ ] Migrated call sites: `phase-b-reviewer` Greptile-budget block, `continue` Step 6, `pm` resume mode
- [ ] Each migrated call site verified (PR description shows manual smoke test against a real PR/branch)
- [ ] Audit doc row updated to mark this row as ✅ extracted, with link to the merge commit

### PR DESCRIPTION
**Pull request by auerbachb**
## Summary

Extracts the canonical read/write helper for `~/.claude/session-state.json` into `.claude/scripts/session-state.sh`. Closes [#284](https://github.com/auerbachb/claude-code-config/issues/284) — the **C-11** P1 row from the script-extraction audit.

The script supports:

- **`--get <jq-path>`** — raw read via `jq -r`. Exits 3 if state file missing, 4 on jq parse error.
- **`--set <jq-path>=<value> [--set ...]`** — surgical write. Multiple `--set` flags accumulate into ONE atomic jq pipeline → ONE temp file → ONE `mv`. Sibling fields preserved verbatim. `.last_updated` auto-refreshed to UTC ISO 8601 on every write.
- **JSON-or-string value detection** — uses `jq empty` (NOT `jq -e .`) to validate parse only, so legitimate JSON `null` and `false` aren't silently coerced to the strings `"null"`/`"false"`. (Caught during smoke testing.)
- **Refuses to silently overwrite a corrupt-but-existing JSON file** (exit 4) — prevents data loss when the state file is mid-corruption from another writer.
- **Atomic write pattern** matches `repair-trust-all.sh` and `greptile-budget.sh` — temp file in same directory + `mv` (POSIX same-FS rename atomicity).

## Migration scope reality

The issue's AC listed three "migrate" sites. Survey found two of them already covered upstream:

| Originally listed call site | Status | Where it actually got migrated |
|---|---|---|
| `phase-b-reviewer` Greptile-budget block (×2 inline jq) | already extracted | C-12 [#285](https://github.com/auerbachb/claude-code-config/pull/285) → delegates to `greptile-budget.sh --consume` |
| `continue` Step 6 rate-limit branch | already extracted | C-14 [#287](https://github.com/auerbachb/claude-code-config/pull/287) → delegates to `reviewer-of.sh --sticky` |
| `pm` resume mode | **migrated here** | `pm/SKILL.md:85` → `session-state.sh --get .` |

Bonus site found in survey:

| Newly discovered call site | Status |
|---|---|
| `pm-handoff` Step 5b (same `cat ~/.claude/session-state.json` pattern as pm) | **migrated here** → `pm-handoff/SKILL.md:222` → `session-state.sh --get .` |

The script still earns its keep as the canonical handle for ad-hoc inspection AND any future write call sites — the multi-set-as-one-atomic-write contract makes it the right primitive when a future skill needs to touch ≥2 paths in one transaction.

## Local CR review

The CodeRabbit CLI was rate-limited (`Rate limit exceeded, please try after 30 minutes`) on the first attempt. Per `cr-local-review.md` Timeout & fallback rules, fell back to a thorough self-review (script structure, trap chains, atomicity, edge cases, migration semantic equivalence) — all clean. Rule-file budget verified: 13,009 / 14,000 words total, `handoff-files.md` 69 lines (200 cap) / 666 words (2,000 cap).

## Test plan

- [ ] `--help` prints the doc-comment block (PURPOSE through EXAMPLES)
- [ ] `--get` on missing state file exits 3 with one-line stderr
- [ ] `--get` on valid state file prints the value via `jq -r` and exits 0
- [ ] `--get` on corrupt state file exits 4 with the underlying jq error in stderr
- [ ] `--set` on missing state file seeds `{}`, applies the assignment, exits 0
- [ ] `--set` on valid state file preserves ALL sibling fields including unknown nested objects + arrays
- [ ] Multiple `--set` flags collapse into ONE jq pipeline + ONE atomic mv
- [ ] JSON-typed values auto-detected: numbers, booleans, **`null`**, **`false`**, objects, arrays
- [ ] Bare strings fall through to `--arg` (e.g. `--set '.label=hello world'`)
- [ ] `--set` refuses to overwrite a corrupt-but-existing JSON file (exit 4, file untouched)
- [ ] Malformed `--set foo` (missing `=`) exits 2 with usage error
- [ ] Unknown flag exits 2; no mode given exits 2; `--get` + `--set` together exits 2
- [ ] Values containing `=` (URLs, tokens) split on the FIRST `=` only
- [ ] `.last_updated` is refreshed to UTC ISO 8601 on every successful `--set`
- [ ] `pm/SKILL.md` Step 1A.2 invocation works for both missing and present state file (NO_SESSION_STATE fallback preserved)
- [ ] `pm-handoff/SKILL.md` Step 5b invocation works for both missing and present state file
- [ ] `script-extraction-audit.md` C-11 row marked ✅ with explanation of stale-AC reality
- [ ] `handoff-files.md` "Session-State Schema" section cross-references the new script
- [ ] CLAUDE.md + rules total word count ≤ 14,000

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a session-state CLI for atomic JSON reads/writes, batched multi-field updates, JSON-literal detection, clear exit codes, and automatic last-updated timestamps.

* **Documentation**
  * Marked the extraction audit entry complete and updated handoff guidance to require using the session-state utility for targeted updates.

* **User Experience**
  * Flows now load session state via the centralized utility while preserving the previous fallback behavior.

### CODE REVIEW

**Comment by auerbachb** [comment]
@coderabbitai Fixed in `1cba584`: switched the file-existence guard from `jq -e .` to `jq empty` (line 215). Same fix as already applied inside the per-`--set` JSON-or-string detection. Smoke-tested: `echo null > state.json; ./session-state.sh --set .x=y` now succeeds (previously misclassified as corrupt → exit 4).

**Comment by auerbachb** [comment]
BugBot fixes in `2ed5398`:

- **#1 (Medium, empty value)** — empty `--set <path>=` now short-circuits to the `--arg` (string) path. `./session-state.sh --set .empty=` produces `{"empty": ""}` instead of erroring with "invalid JSON text passed to --argjson".
- **#2 (Low, state-file shape)** — added `jq -e 'type == "object"'` guard with a type-specific error message (e.g., "top-level is array, expected object"). Matches `reviewer-of.sh` discipline.

Regression-tested: `--set .n=null --set .f=false --set .num=42 --set .str=plain` still type-detect correctly.

**Comment by auerbachb** [comment]
@coderabbitai Fixed in `af77455`: write-stage jq pipeline failure now returns exit 5 per contract. Smoke-tested all exit codes — 0 (happy), 4 (read-stage parse / corrupt / wrong shape / --get parse), 5 (write-stage jq failure).

File: .claude/scripts/session-state.sh [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+      fi
+      MODE="set"
+      local_arg="$2"
+      # Split on the FIRST `=` only — values may contain `=` (e.g., a JSON
+      # string with `=` inside it).
+      if [[ "$local_arg" != *=* ]]; then
+        die_usage "--set argument must be <jq-path>=<value>, got: $local_arg"
+      fi
+      SET_PATHS+=("${local_arg%%=*}")
+      SET_VALUES+=("${local_arg#*=}")
```
[the remark this answers is not in the document]
Comment (auerbachb) [reply .claude/scripts/session-state.sh:166]: Fixed in `66f622e`: added explicit non-empty LHS guard after the `*=*` check — `--set =foo` now exits 2 with the documented usage-error message instead of falling through to jq exit 5.

File: .claude/scripts/session-state.sh [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+    exit 3
+  fi
+  # Use jq -r so callers get the raw value (string without quotes, number
+  # as-is, etc.). jq exits non-zero on parse errors — translate to 4.
+  jq_err="$(mktemp)"
+  trap "rm -f '$jq_err' 2>/dev/null" EXIT
+  if ! jq -r "$GET_PATH" "$STATE_FILE" 2>"$jq_err"; then
+    echo "session-state.sh: jq failed reading $STATE_FILE: $(cat "$jq_err")" >&2
+    exit 4
+  fi
```
[the remark this answers is not in the document]
Comment (auerbachb) [reply .claude/scripts/session-state.sh:219]: Fixed in `66f622e`: introduced `is_single_object_state_file()` helper using `jq -s -e 'length == 1 and (.[0] | type == "object")'` and applied it symmetrically in both `--get` and `--set` paths. Smoke-tested: multi-doc `{}\n{}` files now correctly fail with exit 4 and are NOT corrupted.

File: .claude/scripts/session-state.sh [hunk 3]
```diff
@@ -0,0 +1,307 @@
...
+#   # Set a number (auto-detected as numeric):
+#   session-state.sh --set '.cr_quota.reviews_used=5'
+
+set -euo pipefail
+
+STATE_FILE="${HOME}/.claude/session-state.json"
+
+print_help() {
+  sed -n '/^# PURPOSE$/,/^# EXAMPLES$/p' "$0" | sed 's/^# \{0,1\}//'
+}
```
[the remark this answers is not in the document]
Comment (auerbachb) [reply .claude/scripts/session-state.sh:97]: Fixed in `1474fb0`: replaced `sed -n '/^# PURPOSE$/,/^# EXAMPLES$/p'` with `sed -n '/^# PURPOSE$/,/^$/p'` plus `sed '$d; ...'` to drop the trailing blank. Anchoring on the first blank line marks end-of-comment-block without hard-coding any specific marker beyond it. Verified `--help` now prints the full EXAMPLES section.

**Comment by auerbachb** [comment]
@cursor review

**Comment by auerbachb** [comment]
@coderabbitai full review

**Comment by auerbachb** [comment]
@coderabbitai full review

**Comment by auerbachb** [comment]
@coderabbitai full review
@cursor review
